app/javascript/hyacinth_ui_v1/components/projects/permissions/PermissionsEditor.js

```diff
@@ -0,0 +1,300 @@
+import React, { useState } from 'react';
+import PropTypes from 'prop-types';
+import { useQuery, useMutation } from '@apollo/react-hooks';
+import { Button, Form, Table } from 'react-bootstrap';
+import produce from 'immer';
+import Select from 'react-select';
+import { remove as arrRemove, sortBy as collectionSortBy } from 'lodash';
+import { useHistory } from 'react-router-dom';
+
+import GraphQLErrors from '../../ui/GraphQLErrors';
+import {
+  getProjectPermissionActionsQuery, getProjectPermissionsQuery, updateProjectPermissionsMutation,
+} from '../../../graphql/projects';
+import { getUsersQuery } from '../../../graphql/users';
+import AddButton from '../../ui/buttons/AddButton';
+import RemoveButton from '../../ui/buttons/RemoveButton';
+import CancelButton from '../../ui/forms/buttons/CancelButton';
+
+function PermissionsEditor(props) {
+  const { readOnly } = props;
+  const { projectStringKey } = props;
+  const history = useHistory();
+  const [permissionsChanged, setPermissionsChanged] = useState(false);
+  const [removedUserIds, setRemovedUserIds] = useState(new Set());
+  const [allUsers, setAllUsers] = useState([]);
+  const [projectPermissions, setProjectPermissions] = useState([]);
+  const [currentAddUserSelection, setCurrentAddUserSelection] = useState(null);
+
+  const { loading: actionsLoading, error: actionsError, data: actionsData } = useQuery(
+    getProjectPermissionActionsQuery, { variables: { stringKey: projectStringKey } },
+  );
+  const { loading: usersLoading, error: usersError } = useQuery(
+    getUsersQuery, {
+      onCompleted: (data) => {
+        setAllUsers(data.users);
+      },
+    },
+  );
+  const { loading: permissionsLoading, error: permissionsError } = useQuery(
+    getProjectPermissionsQuery, {
+      variables: { stringKey: projectStringKey },
+      onCompleted: (data) => {
+        setProjectPermissions(
+          collectionSortBy(data.projectPermissionsForProject, [perm => perm.user.sortName])
+        );
+      },
+    },
+  );
+
+  const [updateProjectPermissions, { error: updateProjectPermissionsError }] = useMutation(
+    updateProjectPermissionsMutation,
+  );
+
+  if (actionsLoading || permissionsLoading || usersLoading) return (<></>);
+  if (actionsError || permissionsError || usersError) {
+    return (
+      <GraphQLErrors errors={
+        actionsError || permissionsError || usersError || updateProjectPermissionsError
+      } />
+    );
+  }
+
+  const actionToDisplayLabel = (action) => {
+    return action.split('_').map(part => part.charAt(0).toUpperCase() + part.slice(1)).join(' ');
+  };
+
+  const updatePermissionsData = (userId, action, enabled) => {
+    setPermissionsChanged(true);
+
+    const { projectPermissionActions: {
+      actions: allActions, manageAction, actionsDisallowedForAggregatorProjects },
+    } = actionsData;
+
+    setProjectPermissions(
```

Comment:
I haven't created any custom hooks either, so that would be helpful -- thanks!

Why is it bad to use immer in a function generated by a hook?  And what's the difference between that and using immer with a function that isn't generated by a hook?